File: crypto_3fa.py
#!/usr/bin/env python3
"""
yourl.cloud - 3FA Encryption System for Social Proximity Influence Scores
==========================================================================

Three-Factor Authentication (3FA) implementation with:
- Factor 1: Password/Passphrase (something you know)
- Factor 2: OTP/Hardware Token (something you have)  
- Factor 3: 256-bit Random Key (something you possess)

Session ID: f1d78acb-de07-46e0-bfa7-f5b75e3c0c49

References:
- AES-256-GCM encryption standard
- PKI-based authorization
- Social proximity influence scoring
- Three-factor authentication best practices
"""

import os
import hashlib
import hmac
import base64
import json
import time
from datetime import datetime, timedelta
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.backends import default_backend
import secrets

class ThreeFactorAuth:
    """Three-Factor Authentication system for yourl.cloud"""
    
    def __init__(self):
        self.session_id = "f1d78acb-de07-46e0-bfa7-f5b75e3c0c49"
        self.key_size = 32  # 256 bits
        self.nonce_size = 12  # 96 bits for AES-GCM
        self.salt_size = 16  # 128 bits for key derivation
        
    def generate_256bit_key(self) -> bytes:
        """Generate a cryptographically secure 256-bit key (Factor 3)"""
        return os.urandom(self.key_size)
    
    def derive_master_key(self, password: str, otp: str, secret_key: bytes) -> tuple[bytes, bytes]:
        """
        Derive master key from three factors using PBKDF2
        
        Args:
            password: User password/passphrase (Factor 1)
            otp: One-time password/token (Factor 2) 
            secret_key: 256-bit random key (Factor 3)
            
        Returns:
            Tuple of (256-bit master key, salt) for encryption/decryption
        """
        # Combine all three factors
        combined = password.encode('utf-8') + otp.encode('utf-8') + secret_key
        
        # Use PBKDF2 for key derivation with high iteration count
        salt = os.urandom(self.salt_size)
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=self.key_size,
            salt=salt,
            iterations=100000,  # High iteration count for security
            backend=default_backend()
        )
        
        master_key = kdf.derive(combined)
        return master_key, salt
    
    def verify_master_key(self, password: str, otp: str, secret_key: bytes, 
                         stored_salt: bytes, stored_key_hash: bytes) -> bool:
        """Verify the master key against stored hash"""
        try:
            combined = password.encode('utf-8') + otp.encode('utf-8') + secret_key
            kdf = PBKDF2HMAC(
                algorithm=hashes.SHA256(),
                length=self.key_size,
                salt=stored_salt,
                iterations=100000,
                backend=default_backend()
            )
            
            derived_key = kdf.derive(combined)
            key_hash = hashlib.sha256(derived_key).digest()
            
            return hmac.compare_digest(key_hash, stored_key_hash)
        except Exception:
            return False

class InfluenceScoreEncryption:
    """Encryption system for social proximity influence scores"""
    
    def __init__(self, master_key: bytes):
        self.master_key = master_key
        self.aesgcm = AESGCM(master_key)
        
    def encrypt_score(self, score: float, metadata: dict | None = None) -> dict:
        """
        Encrypt a social proximity influence score using AES-256-GCM
        
        Args:
            score: Influence score (float)
            metadata: Additional metadata (dict or None)
            
        Returns:
            Encrypted score data with nonce and metadata
        """
        # Prepare score data
        score_data = {
            'score': score,
            'timestamp': datetime.utcnow().isoformat(),
            'session_id': "f1d78acb-de07-46e0-bfa7-f5b75e3c0c49",
            'metadata': metadata if metadata is not None else {}
        }
        
        # Convert to JSON and encode
        score_bytes = json.dumps(score_data, sort_keys=True).encode('utf-8')
        
        # Generate nonce for AES-GCM
        nonce = os.urandom(12)
        
        # Encrypt the score data
        encrypted_data = self.aesgcm.encrypt(nonce, score_bytes, None)
        
        # Return encrypted data with nonce
        return {
            'encrypted_data': base64.b64encode(encrypted_data).decode('utf-8'),
            'nonce': base64.b64encode(nonce).decode('utf-8'),
            'algorithm': 'AES-256-GCM',
            'key_size': 256,
            'timestamp': datetime.utcnow().isoformat(),
            'version': '1.0'
        }
    
    def decrypt_score(self, encrypted_data: dict) -> dict:
        """
        Decrypt a social proximity influence score
        
        Args:
            encrypted_data: Encrypted score data from encrypt_score()
            
        Returns:
            Decrypted score data
        """
        try:
            # Decode base64 data
            encrypted_bytes = base64.b64decode(encrypted_data['encrypted_data'])
            nonce = base64.b64decode(encrypted_data['nonce'])
            
            # Decrypt the data
            decrypted_bytes = self.aesgcm.decrypt(nonce, encrypted_bytes, None)
            
            # Parse JSON data
            score_data = json.loads(decrypted_bytes.decode('utf-8'))
            
            return score_data
            
        except Exception as e:
            raise ValueError(f"Failed to decrypt score: {str(e)}")

class PKIAuthorization:
    """Public Key Infrastructure for score authorization"""
    
    def __init__(self):
        self.backend = default_backend()
        
    def generate_key_pair(self) -> tuple:
        """Generate RSA key pair for PKI"""
        private_key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=2048,
            backend=self.backend
        )
        public_key = private_key.public_key()
        
        return private_key, public_key
    
    def serialize_public_key(self, public_key) -> str:
        """Serialize public key to PEM format"""
        pem = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )
        return pem.decode('utf-8')
    
    def sign_score(self, private_key, score_data: dict) -> str:
        """Sign score data with private key"""
        score_bytes = json.dumps(score_data, sort_keys=True).encode('utf-8')
        signature = private_key.sign(
            score_bytes,
            padding.PSS(
                mgf=padding.MGF1(hashes.SHA256()),
                salt_length=padding.PSS.MAX_LENGTH
            ),
            hashes.SHA256()
        )
        return base64.b64encode(signature).decode('utf-8')
    
    def verify_score(self, public_key, score_data: dict, signature: str) -> bool:
        """Verify score signature with public key"""
        try:
            score_bytes = json.dumps(score_data, sort_keys=True).encode('utf-8')
            signature_bytes = base64.b64decode(signature)
            
            public_key.verify(
                signature_bytes,
                score_bytes,
                padding.PSS(
                    mgf=padding.MGF1(hashes.SHA256()),
                    salt_length=padding.PSS.MAX_LENGTH
                ),
                hashes.SHA256()
            )
            return True
        except Exception:
            return False

class SocialProximityScoring:
    """Social proximity influence score calculation and management"""
    
    def __init__(self):
        self.session_id = "f1d78acb-de07-46e0-bfa7-f5b75e3c0c49"
        
    def calculate_influence_score(self, user_data: dict) -> float:
        """
        Calculate social proximity influence score based on user data
        
        Args:
            user_data: Dictionary containing user metrics
            
        Returns:
            Influence score (0.0 to 1.0)
        """
        # Example scoring algorithm - customize based on your needs
        score = 0.0
        
        # Network size factor (0-30%)
        network_size = user_data.get('network_size', 0)
        max_network = user_data.get('max_network_size', 10000)
        network_factor = min(network_size / max_network, 1.0) * 0.3
        
        # Engagement factor (0-25%)
        engagement_rate = user_data.get('engagement_rate', 0.0)
        engagement_factor = min(engagement_rate, 1.0) * 0.25
        
        # Trust factor (0-25%)
        trust_score = user_data.get('trust_score', 0.0)
        trust_factor = min(trust_score, 1.0) * 0.25
        
        # Activity factor (0-20%)
        activity_level = user_data.get('activity_level', 0.0)
        activity_factor = min(activity_level, 1.0) * 0.20
        
        score = network_factor + engagement_factor + trust_factor + activity_factor
        
        # Ensure score is between 0.0 and 1.0
        return max(0.0, min(1.0, score))
    
    def validate_score(self, score: float) -> bool:
        """Validate that score is within acceptable range"""
        return 0.0 <= score <= 1.0

# Example usage and testing
def main():
    """Example usage of the 3FA encryption system"""
    print("yourl.cloud - 3FA Encryption System Demo")
    print("=" * 50)
    print(f"Session ID: f1d78acb-de07-46e0-bfa7-f5b75e3c0c49")
    print()
    
    # Initialize 3FA system
    auth = ThreeFactorAuth()
    scoring = SocialProximityScoring()
    pki = PKIAuthorization()
    
    # Generate 256-bit key (Factor 3)
    secret_key = auth.generate_256bit_key()
    print(f"Generated 256-bit key: {base64.b64encode(secret_key).decode('utf-8')}")
    
    # Example user factors
    password = "secure_password_123"  # Factor 1
    otp = "123456"  # Factor 2 (in real implementation, this would be from TOTP device)
    
    # Derive master key
    master_key, salt = auth.derive_master_key(password, otp, secret_key)
    print(f"Derived master key: {base64.b64encode(master_key).decode('utf-8')}")
    
    # Calculate influence score
    user_data = {
        'network_size': 5000,
        'max_network_size': 10000,
        'engagement_rate': 0.75,
        'trust_score': 0.85,
        'activity_level': 0.90
    }
    
    score = scoring.calculate_influence_score(user_data)
    print(f"Calculated influence score: {score:.4f}")
    
    # Encrypt the score
    encryption = InfluenceScoreEncryption(master_key)
    encrypted_score = encryption.encrypt_score(score, user_data)
    print(f"Encrypted score data: {json.dumps(encrypted_score, indent=2)}")
    
    # Decrypt the score
    decrypted_score = encryption.decrypt_score(encrypted_score)
    print(f"Decrypted score: {decrypted_score['score']:.4f}")
    
    # PKI signing example
    private_key, public_key = pki.generate_key_pair()
    signature = pki.sign_score(private_key, encrypted_score)
    print(f"Score signature: {signature}")
    
    # Verify signature
    is_valid = pki.verify_score(public_key, encrypted_score, signature)
    print(f"Signature valid: {is_valid}")
    
    print("\n3FA encryption system ready for yourl.cloud!")

if __name__ == "__main__":
    main()

File: public/crypto_3fa.js
/**
 * yourl.cloud - 3FA Encryption System (JavaScript)
 * ================================================
 * 
 * Three-Factor Authentication (3FA) implementation with:
 * - Factor 1: Password/Passphrase (something you know)
 * - Factor 2: OTP/Hardware Token (something you have)  
 * - Factor 3: 256-bit Random Key (something you possess)
 * 
 * Session ID: f1d78acb-de07-46e0-bfa7-f5b75e3c0c49
 * 
 * References:
 * - Web Crypto API for AES-256-GCM encryption
 * - SubtleCrypto for key generation and derivation
 * - PKI-based authorization
 * - Social proximity influence scoring
 */

class ThreeFactorAuth {
    constructor() {
        this.sessionId = "f1d78acb-de07-46e0-bfa7-f5b75e3c0c49";
        this.keySize = 32; // 256 bits
        this.nonceSize = 12; // 96 bits for AES-GCM
        this.saltSize = 16; // 128 bits for key derivation
    }

    /**
     * Generate a cryptographically secure 256-bit key (Factor 3)
     */
    async generate256bitKey() {
        return await crypto.subtle.generateKey(
            {
                name: "AES-GCM",
                length: 256
            },
            true,
            ["encrypt", "decrypt"]
        );
    }

    /**
     * Derive master key from three factors using PBKDF2
     * 
     * @param {string} password - User password/passphrase (Factor 1)
     * @param {string} otp - One-time password/token (Factor 2) 
     * @param {CryptoKey} secretKey - 256-bit random key (Factor 3)
     * @returns {Promise<{masterKey: CryptoKey, salt: Uint8Array}>}
     */
    async deriveMasterKey(password, otp, secretKey) {
        // Combine all three factors
        const combined = new TextEncoder().encode(password + otp);
        
        // Generate salt for key derivation
        const salt = crypto.getRandomValues(new Uint8Array(this.saltSize));
        
        // Use PBKDF2 for key derivation with high iteration count
        const baseKey = await crypto.subtle.importKey(
            "raw",
            combined,
            "PBKDF2",
            false,
            ["deriveBits", "deriveKey"]
        );
        
        const masterKey = await crypto.subtle.deriveKey(
            {
                name: "PBKDF2",
                salt: salt,
                iterations: 100000, // High iteration count for security
                hash: "SHA-256"
            },
            baseKey,
            {
                name: "AES-GCM",
                length: 256
            },
            true,
            ["encrypt", "decrypt"]
        );
        
        return { masterKey, salt };
    }

    /**
     * Verify the master key against stored hash
     */
    async verifyMasterKey(password, otp, secretKey, storedSalt, storedKeyHash) {
        try {
            const { masterKey } = await this.deriveMasterKey(password, otp, secretKey);
            const keyBuffer = await crypto.subtle.exportKey("raw", masterKey);
            const keyHash = await crypto.subtle.digest("SHA-256", keyBuffer);
            
            // Compare hashes securely
            const storedHashArray = new Uint8Array(storedKeyHash);
            const computedHashArray = new Uint8Array(keyHash);
            
            if (storedHashArray.length !== computedHashArray.length) {
                return false;
            }
            
            for (let i = 0; i < storedHashArray.length; i++) {
                if (storedHashArray[i] !== computedHashArray[i]) {
                    return false;
                }
            }
            
            return true;
        } catch (error) {
            console.error("Key verification failed:", error);
            return false;
        }
    }
}

class InfluenceScoreEncryption {
    constructor(masterKey) {
        this.masterKey = masterKey;
    }

    /**
     * Encrypt a social proximity influence score using AES-256-GCM
     * 
     * @param {number} score - Influence score (float)
     * @param {Object} metadata - Additional metadata
     * @returns {Promise<Object>} Encrypted score data with nonce and metadata
     */
    async encryptScore(score, metadata = {}) {
        // Prepare score data
        const scoreData = {
            score: score,
            timestamp: new Date().toISOString(),
            sessionId: "f1d78acb-de07-46e0-bfa7-f5b75e3c0c49",
            metadata: metadata
        };
        
        // Convert to JSON and encode
        const scoreBytes = new TextEncoder().encode(JSON.stringify(scoreData));
        
        // Generate nonce for AES-GCM
        const nonce = crypto.getRandomValues(new Uint8Array(12));
        
        // Encrypt the score data
        const encryptedData = await crypto.subtle.encrypt(
            {
                name: "AES-GCM",
                iv: nonce
            },
            this.masterKey,
            scoreBytes
        );
        
        // Return encrypted data with nonce
        return {
            encryptedData: btoa(String.fromCharCode(...new Uint8Array(encryptedData))),
            nonce: btoa(String.fromCharCode(...nonce)),
            algorithm: "AES-256-GCM",
            keySize: 256,
            timestamp: new Date().toISOString(),
            version: "1.0"
        };
    }

    /**
     * Decrypt a social proximity influence score
     * 
     * @param {Object} encryptedData - Encrypted score data from encryptScore()
     * @returns {Promise<Object>} Decrypted score data
     */
    async decryptScore(encryptedData) {
        try {
            // Decode base64 data
            const encryptedBytes = new Uint8Array(
                atob(encryptedData.encryptedData).split('').map(char => char.charCodeAt(0))
            );
            const nonce = new Uint8Array(
                atob(encryptedData.nonce).split('').map(char => char.charCodeAt(0))
            );
            
            // Decrypt the data
            const decryptedBytes = await crypto.subtle.decrypt(
                {
                    name: "AES-GCM",
                    iv: nonce
                },
                this.masterKey,
                encryptedBytes
            );
            
            // Parse JSON data
            const scoreData = JSON.parse(new TextDecoder().decode(decryptedBytes));
            
            return scoreData;
            
        } catch (error) {
            throw new Error(`Failed to decrypt score: ${error.message}`);
        }
    }
}

class SocialProximityScoring {
    constructor() {
        this.sessionId = "f1d78acb-de07-46e0-bfa7-f5b75e3c0c49";
    }

    /**
     * Calculate social proximity influence score based on user data
     * 
     * @param {Object} userData - Dictionary containing user metrics
     * @returns {number} Influence score (0.0 to 1.0)
     */
    calculateInfluenceScore(userData) {
        // Example scoring algorithm - customize based on your needs
        let score = 0.0;
        
        // Network size factor (0-30%)
        const networkSize = userData.networkSize || 0;
        const maxNetwork = userData.maxNetworkSize || 10000;
        const networkFactor = Math.min(networkSize / maxNetwork, 1.0) * 0.3;
        
        // Engagement factor (0-25%)
        const engagementRate = userData.engagementRate || 0.0;
        const engagementFactor = Math.min(engagementRate, 1.0) * 0.25;
        
        // Trust factor (0-25%)
        const trustScore = userData.trustScore || 0.0;
        const trustFactor = Math.min(trustScore, 1.0) * 0.25;
        
        // Activity factor (0-20%)
        const activityLevel = userData.activityLevel || 0.0;
        const activityFactor = Math.min(activityLevel, 1.0) * 0.20;
        
        score = networkFactor + engagementFactor + trustFactor + activityFactor;
        
        // Ensure score is between 0.0 and 1.0
        return Math.max(0.0, Math.min(1.0, score));
    }

    /**
     * Validate that score is within acceptable range
     */
    validateScore(score) {
        return score >= 0.0 && score <= 1.0;
    }
}

class PKIAuthorization {
    constructor() {
        this.algorithm = {
            name: "RSASSA-PKCS1-v1_5",
            modulusLength: 2048,
            publicExponent: new Uint8Array([1, 0, 1]),
            hash: "SHA-256"
        };
    }

    /**
     * Generate RSA key pair for PKI
     */
    async generateKeyPair() {
        return await crypto.subtle.generateKey(
            this.algorithm,
            true,
            ["sign", "verify"]
        );
    }

    /**
     * Serialize public key to PEM format
     */
    async serializePublicKey(publicKey) {
        const exported = await crypto.subtle.exportKey("spki", publicKey);
        const base64 = btoa(String.fromCharCode(...new Uint8Array(exported)));
        return `-----BEGIN PUBLIC KEY-----\n${base64}\n-----END PUBLIC KEY-----`;
    }

    /**
     * Sign score data with private key
     */
    async signScore(privateKey, scoreData) {
        const scoreBytes = new TextEncoder().encode(JSON.stringify(scoreData));
        const signature = await crypto.subtle.sign(
            this.algorithm,
            privateKey,
            scoreBytes
        );
        return btoa(String.fromCharCode(...new Uint8Array(signature)));
    }

    /**
     * Verify score signature with public key
     */
    async verifyScore(publicKey, scoreData, signature) {
        try {
            const scoreBytes = new TextEncoder().encode(JSON.stringify(scoreData));
            const signatureBytes = new Uint8Array(
                atob(signature).split('').map(char => char.charCodeAt(0))
            );
            
            return await crypto.subtle.verify(
                this.algorithm,
                publicKey,
                signatureBytes,
                scoreBytes
            );
        } catch (error) {
            console.error("Signature verification failed:", error);
            return false;
        }
    }
}

// Example usage and testing
async function demo3FA() {
    console.log("yourl.cloud - 3FA Encryption System Demo");
    console.log("=" .repeat(50));
    console.log(`Session ID: f1d78acb-de07-46e0-bfa7-f5b75e3c0c49`);
    console.log();
    
    try {
        // Initialize 3FA system
        const auth = new ThreeFactorAuth();
        const scoring = new SocialProximityScoring();
        const pki = new PKIAuthorization();
        
        // Generate 256-bit key (Factor 3)
        const secretKey = await auth.generate256bitKey();
        console.log("Generated 256-bit key");
        
        // Example user factors
        const password = "secure_password_123"; // Factor 1
        const otp = "123456"; // Factor 2 (in real implementation, this would be from TOTP device)
        
        // Derive master key
        const { masterKey, salt } = await auth.deriveMasterKey(password, otp, secretKey);
        console.log("Derived master key");
        
        // Calculate influence score
        const userData = {
            networkSize: 5000,
            maxNetworkSize: 10000,
            engagementRate: 0.75,
            trustScore: 0.85,
            activityLevel: 0.90
        };
        
        const score = scoring.calculateInfluenceScore(userData);
        console.log(`Calculated influence score: ${score.toFixed(4)}`);
        
        // Encrypt the score
        const encryption = new InfluenceScoreEncryption(masterKey);
        const encryptedScore = await encryption.encryptScore(score, userData);
        console.log("Encrypted score data:", encryptedScore);
        
        // Decrypt the score
        const decryptedScore = await encryption.decryptScore(encryptedScore);
        console.log(`Decrypted score: ${decryptedScore.score.toFixed(4)}`);
        
        // PKI signing example
        const { privateKey, publicKey } = await pki.generateKeyPair();
        const signature = await pki.signScore(privateKey, encryptedScore);
        console.log("Score signature:", signature);
        
        // Verify signature
        const isValid = await pki.verifyScore(publicKey, encryptedScore, signature);
        console.log(`Signature valid: ${isValid}`);
        
        console.log("\n3FA encryption system ready for yourl.cloud!");
        
    } catch (error) {
        console.error("Demo failed:", error);
    }
}

// Export classes for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        ThreeFactorAuth,
        InfluenceScoreEncryption,
        SocialProximityScoring,
        PKIAuthorization,
        demo3FA
    };
} else {
    // Browser environment - attach to window
    window.ThreeFactorAuth = ThreeFactorAuth;
    window.InfluenceScoreEncryption = InfluenceScoreEncryption;
    window.SocialProximityScoring = SocialProximityScoring;
    window.PKIAuthorization = PKIAuthorization;
    window.demo3FA = demo3FA;
}
